File: server/src/collectors/nraRadiation.js
/**
 * NRA Radiation Monitoring Collector
 * Nuclear Regulation Authority — 全国の空間線量率モニタリングポスト.
 *
 * Live cascade:
 *   1. NRA radioactivity public JSON feeds (multiple known endpoints)
 *   2. JCAC (放射線監視センター) station list
 *   3. OSM `man_made=monitoring_station` + `monitoring:radiation=yes`
 *   4. Static seed of major monitoring posts (clearly marked)
 */

import { fetchJson, fetchOverpass } from './_liveHelpers.js';

const NRA_FEEDS = [
  'https://radioactivity.nra.go.jp/cont/json/dose_today.json',
  'https://radioactivity.nra.go.jp/cont/json/MP/MP_today.json',
  'https://www.kankyo-hoshano.go.jp/api/now_value.json',
  'https://emdb.jaea.go.jp/emdb/api/v1/monitoring_post/latest.json',
];

const STATIONS = [
  { id: 'NRA001', name: '福島市杉妻町', pref: '福島県', lat: 37.750, lon: 140.468, baseline: 120 },
  { id: 'NRA002', name: '郡山合同庁舎', pref: '福島県', lat: 37.400, lon: 140.360, baseline: 95 },
  { id: 'NRA003', name: 'いわき合同庁舎', pref: '福島県', lat: 36.945, lon: 140.888, baseline: 85 },
  { id: 'NRA004', name: '南相馬市原町区', pref: '福島県', lat: 37.642, lon: 140.957, baseline: 150 },
  { id: 'NRA005', name: '双葉町中央', pref: '福島県', lat: 37.451, lon: 141.009, baseline: 350 },
  { id: 'NRA006', name: '大熊町大川原', pref: '福島県', lat: 37.394, lon: 140.979, baseline: 420 },
  { id: 'NRA007', name: '浪江町幾世橋', pref: '福島県', lat: 37.495, lon: 141.001, baseline: 280 },
  { id: 'NRA008', name: '富岡町小浜', pref: '福島県', lat: 37.342, lon: 141.010, baseline: 220 },
  { id: 'NRA009', name: '飯舘村役場', pref: '福島県', lat: 37.680, lon: 140.737, baseline: 180 },
  { id: 'NRA010', name: '田村市常葉', pref: '福島県', lat: 37.433, lon: 140.574, baseline: 75 },
  { id: 'NRA011', name: '札幌', pref: '北海道', lat: 43.064, lon: 141.347, baseline: 25 },
  { id: 'NRA012', name: '青森', pref: '青森県', lat: 40.824, lon: 140.740, baseline: 30 },
  { id: 'NRA013', name: '盛岡', pref: '岩手県', lat: 39.704, lon: 141.153, baseline: 28 },
  { id: 'NRA014', name: '仙台', pref: '宮城県', lat: 38.269, lon: 140.872, baseline: 45 },
  { id: 'NRA015', name: '秋田', pref: '秋田県', lat: 39.720, lon: 140.103, baseline: 30 },
  { id: 'NRA016', name: '山形', pref: '山形県', lat: 38.241, lon: 140.364, baseline: 40 },
  { id: 'NRA017', name: '水戸', pref: '茨城県', lat: 36.342, lon: 140.447, baseline: 55 },
  { id: 'NRA018', name: '宇都宮', pref: '栃木県', lat: 36.566, lon: 139.884, baseline: 50 },
  { id: 'NRA019', name: '前橋', pref: '群馬県', lat: 36.391, lon: 139.061, baseline: 30 },
  { id: 'NRA020', name: 'さいたま', pref: '埼玉県', lat: 35.857, lon: 139.649, baseline: 35 },
  { id: 'NRA021', name: '千葉', pref: '千葉県', lat: 35.605, lon: 140.123, baseline: 38 },
  { id: 'NRA022', name: '新宿', pref: '東京都', lat: 35.694, lon: 139.703, baseline: 35 },
  { id: 'NRA023', name: '横浜', pref: '神奈川県', lat: 35.448, lon: 139.642, baseline: 32 },
  { id: 'NRA024', name: '新潟', pref: '新潟県', lat: 37.902, lon: 139.023, baseline: 35 },
  { id: 'NRA025', name: '金沢', pref: '石川県', lat: 36.594, lon: 136.626, baseline: 40 },
  { id: 'NRA026', name: '名古屋', pref: '愛知県', lat: 35.180, lon: 136.907, baseline: 35 },
  { id: 'NRA027', name: '京都', pref: '京都府', lat: 35.021, lon: 135.756, baseline: 40 },
  { id: 'NRA028', name: '大阪', pref: '大阪府', lat: 34.686, lon: 135.520, baseline: 42 },
  { id: 'NRA029', name: '神戸', pref: '兵庫県', lat: 34.691, lon: 135.183, baseline: 40 },
  { id: 'NRA030', name: '岡山', pref: '岡山県', lat: 34.662, lon: 133.935, baseline: 45 },
  { id: 'NRA031', name: '広島', pref: '広島県', lat: 34.396, lon: 132.460, baseline: 48 },
  { id: 'NRA032', name: '松山', pref: '愛媛県', lat: 33.842, lon: 132.766, baseline: 45 },
  { id: 'NRA033', name: '福岡', pref: '福岡県', lat: 33.607, lon: 130.418, baseline: 40 },
  { id: 'NRA034', name: '鹿児島', pref: '鹿児島県', lat: 31.560, lon: 130.558, baseline: 35 },
  { id: 'NRA035', name: '那覇', pref: '沖縄県', lat: 26.335, lon: 127.681, baseline: 22 },
  { id: 'NRA036', name: '東海村', pref: '茨城県', lat: 36.467, lon: 140.564, baseline: 60 },
  { id: 'NRA037', name: '柏崎市', pref: '新潟県', lat: 37.372, lon: 138.560, baseline: 35 },
  { id: 'NRA038', name: '敦賀市', pref: '福井県', lat: 35.645, lon: 136.055, baseline: 45 },
  { id: 'NRA039', name: '六ヶ所村', pref: '青森県', lat: 40.955, lon: 141.368, baseline: 32 },
  { id: 'NRA040', name: '玄海町', pref: '佐賀県', lat: 33.472, lon: 129.870, baseline: 38 },
];

async function tryNraJson() {
  for (const url of NRA_FEEDS) {
    const data = await fetchJson(url, { timeoutMs: 10_000, retries: 1 });
    if (!data) continue;
    const arr = Array.isArray(data) ? data : (data.stations || data.data || data.results || data.items || []);
    if (!Array.isArray(arr) || arr.length === 0) continue;
    const features = [];
    for (const r of arr) {
      const lat = Number(r?.lat ?? r?.LAT ?? r?.緯度);
      const lon = Number(r?.lon ?? r?.lng ?? r?.LON ?? r?.経度);
      if (!Number.isFinite(lat) || !Number.isFinite(lon)) continue;
      features.push({
        type: 'Feature',
        geometry: { type: 'Point', coordinates: [lon, lat] },
        properties: {
          station_id: r?.id ?? r?.station_id ?? r?.code ?? null,
          station_name: r?.name ?? r?.station_name ?? r?.name_jp ?? null,
          prefecture: r?.pref ?? r?.prefecture ?? null,
          dose_rate_nGyh: r?.value ?? r?.dose_rate ?? r?.gamma ?? null,
          measured_at: r?.datetime ?? r?.measured_at ?? r?.time ?? null,
          source: 'nra_live',
        },
      });
    }
    if (features.length > 0) return features;
  }
  return null;
}

async function tryOsmStations() {
  return fetchOverpass(
    [
      'node["man_made"="monitoring_station"]["monitoring:radiation"="yes"](area.jp);',
      'node["man_made"="monitoring_station"]["monitoring:radioactivity"="yes"](area.jp);',
      'node["operator"~"原子力規制"](area.jp);',
    ].join(''),
    (el, _i, coords) => ({
      type: 'Feature',
      geometry: { type: 'Point', coordinates: coords },
      properties: {
        station_id: `OSM_${el.id}`,
        station_name: el.tags?.['name:en'] || el.tags?.name || 'Radiation monitoring post',
        operator: el.tags?.operator || null,
        ref: el.tags?.ref || null,
        source: 'osm_overpass',
      },
    }),
    60_000,
    { limit: 0, queryTimeout: 120 },
  );
}

function generateSeedData() {
  const now = new Date();
  return STATIONS.map((st) => {
    const variation = (Math.random() - 0.5) * 0.2 * st.baseline;
    const doseRate = Math.round((st.baseline + variation) * 10) / 10;
    let status = 'normal';
    if (doseRate > 500) status = 'elevated';
    else if (doseRate > 200) status = 'attention';
    else if (doseRate > 100) status = 'monitoring';
    return {
      type: 'Feature',
      geometry: { type: 'Point', coordinates: [st.lon, st.lat] },
      properties: {
        station_id: st.id,
        station_name: st.name,
        prefecture: st.pref,
        dose_rate_nGyh: doseRate,
        status,
        measured_at: now.toISOString(),
        source: 'nra_seed',
      },
    };
  });
}

export default async function collectNraRadiation() {
  let features = await tryNraJson();
  let live = !!(features && features.length);
  let liveSrc = live ? 'nra_radioactivity_json' : null;

  if (!live) {
    features = await tryOsmStations();
    live = !!(features && features.length);
    if (live) liveSrc = 'osm_monitoring_station';
  }
  if (!live) features = generateSeedData();

  return {
    type: 'FeatureCollection',
    features,
    _meta: {
      source: live ? liveSrc : 'nra_seed',
      fetchedAt: new Date().toISOString(),
      recordCount: features.length,
      live,
      description: 'Radiation dose rate monitoring posts across Japan',
    },
    metadata: {},
  };
}

File: native/collectors/sources/nra_radiation.c
/* collectors/environment/sources/nra_radiation.c
 * Port of server/src/collectors/nraRadiation.js — PRIMARY path only
 * (tryNraJson over NRA_FEEDS). The OSM-station and curated-seed fallbacks
 * (tryOsmStations / generateSeedData) are intentionally NOT ported (RULE 8:
 * JSON-API primary, drop the Overpass/seed fallback). JSON_API family.
 * Registry id "nra-radiation", category "environment". */
#include "source.h"
#include "lib/feedlib.h"
#include "lib/geojson.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define TIMEOUT_MS 10000

static const char *NRA_FEEDS[] = {
  "https://radioactivity.nra.go.jp/cont/json/dose_today.json",
  "https://radioactivity.nra.go.jp/cont/json/MP/MP_today.json",
  "https://www.kankyo-hoshano.go.jp/api/now_value.json",
  "https://emdb.jaea.go.jp/emdb/api/v1/monitoring_post/latest.json",
};
#define NFEEDS ((int)(sizeof(NRA_FEEDS) / sizeof(NRA_FEEDS[0])))

/* JS: Number(x) — coerce string/number; NaN if not finite. */
static int as_num(const cJSON *v, double *out) {
  if (!v) return 0;
  if (cJSON_IsNumber(v)) { *out = v->valuedouble; return isfinite(*out); }
  if (cJSON_IsString(v) && v->valuestring && v->valuestring[0]) {
    char *end;
    double d = strtod(v->valuestring, &end);
    if (end != v->valuestring && isfinite(d)) { *out = d; return 1; }
  }
  return 0;
}

/* first present of a key list, else NULL (=== JS `a ?? b ?? c`) */
static const cJSON *first(const cJSON *o, const char *const *keys) {
  for (int i = 0; keys[i]; i++) {
    const cJSON *v = cJSON_GetObjectItem(o, keys[i]);
    if (v && !cJSON_IsNull(v)) return v;
  }
  return NULL;
}

static cJSON *dup_or_null(const cJSON *v) {
  return v ? cJSON_Duplicate(v, 1) : cJSON_CreateNull();
}

static int run(const source_ctx *ctx, intel_sink *sink) {
  for (int fi = 0; fi < NFEEDS; fi++) {
    cJSON *data = feed_get_json(ctx->http, NRA_FEEDS[fi], TIMEOUT_MS);
    if (!data) continue;

    /* arr = Array.isArray(data) ? data
     *       : (data.stations || data.data || data.results || data.items || []) */
    const cJSON *arr = NULL;
    if (cJSON_IsArray(data)) {
      arr = data;
    } else {
      static const char *ak[] = { "stations", "data", "results", "items", NULL };
      const cJSON *a = first(data, ak);
      if (a && cJSON_IsArray(a)) arr = a;
    }
    if (!arr || !cJSON_IsArray(arr) || cJSON_GetArraySize(arr) == 0) {
      cJSON_Delete(data);
      continue;
    }

    cJSON *features = cJSON_CreateArray();
    const cJSON *r;
    cJSON_ArrayForEach(r, arr) {
      static const char *latk[] = { "lat", "LAT", "緯度", NULL };
      static const char *lonk[] = { "lon", "lng", "LON", "経度", NULL };
      double lat = 0, lon = 0;
      int glat = as_num(first(r, latk), &lat);
      int glon = as_num(first(r, lonk), &lon);
      int geocoded = glat && glon;

      cJSON *f = cJSON_CreateObject();
      cJSON_AddStringToObject(f, "type", "Feature");
      if (geocoded) {
        cJSON *g = cJSON_CreateObject();
        cJSON_AddStringToObject(g, "type", "Point");
        cJSON *co = cJSON_CreateArray();
        cJSON_AddItemToArray(co, cJSON_CreateNumber(lon));
        cJSON_AddItemToArray(co, cJSON_CreateNumber(lat));
        cJSON_AddItemToObject(g, "coordinates", co);
        cJSON_AddItemToObject(f, "geometry", g);
      } else {
        cJSON_AddItemToObject(f, "geometry", cJSON_CreateNull());
      }

      cJSON *p = cJSON_CreateObject();             /* EXACT JS key order */
      static const char *idk[]  = { "id", "station_id", "code", NULL };
      static const char *nmk[]  = { "name", "station_name", "name_jp", NULL };
      static const char *prk[]  = { "pref", "prefecture", NULL };
      static const char *dsk[]  = { "value", "dose_rate", "gamma", NULL };
      static const char *mtk[]  = { "datetime", "measured_at", "time", NULL };
      cJSON_AddItemToObject(p, "station_id",   dup_or_null(first(r, idk)));
      cJSON_AddItemToObject(p, "station_name", dup_or_null(first(r, nmk)));
      cJSON_AddItemToObject(p, "prefecture",   dup_or_null(first(r, prk)));
      cJSON_AddItemToObject(p, "dose_rate_nGyh", dup_or_null(first(r, dsk)));
      cJSON_AddItemToObject(p, "measured_at",  dup_or_null(first(r, mtk)));
      cJSON_AddStringToObject(p, "source", "nra_live");
      cJSON_AddItemToObject(f, "properties", p);
      cJSON_AddItemToArray(features, f);
    }
    cJSON_Delete(data);

    if (cJSON_GetArraySize(features) > 0) {
      int n = geojson_emit_features(sink, ctx->source_id, features);
      cJSON_Delete(features);
      fprintf(stderr, "[nra-radiation] emitted %d (feed %d)\n", n, fi);
      return n >= 0 ? 0 : -1;
    }
    cJSON_Delete(features);
  }
  fprintf(stderr, "[nra-radiation] no live feed\n");
  return -1;
}

static const source_def nra_radiation_def = {
  .id = "nra-radiation", .collector = "environment",
  .name = "NRA Radiation Monitoring",
  .name_ja = "原子力規制委員会 放射線モニタリング",
   .update_interval_sec = 600, .run = run };
REGISTER_SOURCE(nra_radiation_def)
